File: apps/burner-mac/Tests/FlagshipBurnerTests/VMCoreVectorTests.swift
import XCTest
@testable import FlagshipBurnerCore

/// Drives the pure Swift VM core against the SHARED golden vectors
/// (apps/desktop-shared/golden/vm-core-vectors.json) — the cross-language
/// contract that keeps this Swift core and the C# core in apps/burner-windows
/// identical, the same role engine/golden/preseed-vectors.json plays for the
/// preseed engine. Mirrors apps/burner-windows/tests/VMCoreVectorTests.cs one
/// for one. If a vector fails here the fix is byte-parity with the vectors,
/// never editing the vectors to match the code (unless BOTH platforms change
/// together).
///
/// The file is read straight from the shared location via `#filePath` (not a
/// copied test resource), exactly as the Windows csproj *links* the same file
/// into its Resources — so there is nothing to drift.
final class VMCoreVectorTests: XCTestCase {

    // MARK: - Vector loading

    private static func vectorsURL() -> URL {
        // .../apps/burner-mac/Tests/FlagshipBurnerTests/VMCoreVectorTests.swift
        //   → .../FlagshipBurnerTests → .../Tests → .../burner-mac → .../apps
        var dir = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent() // FlagshipBurnerTests
            .deletingLastPathComponent() // Tests
            .deletingLastPathComponent() // burner-mac
            .deletingLastPathComponent() // apps
        dir.appendPathComponent("desktop-shared/golden/vm-core-vectors.json")
        return dir
    }

    private func loadVectors() throws -> [String: Any] {
        let url = Self.vectorsURL()
        guard FileManager.default.fileExists(atPath: url.path) else {
            throw XCTSkip("shared vm-core-vectors.json not found at \(url.path)")
        }
        let data = try Data(contentsOf: url)
        guard let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw XCTSkip("vm-core-vectors.json is not a JSON object")
        }
        return obj
    }

    // MARK: - Token codecs (mirror the C# ParseState/ParseEvent/ParseEffect)

    private func parseState(_ token: String) throws -> VMState {
        switch token {
        case "created": return .created
        case "installing": return .installing
        case "installed": return .installed
        case "awaitingPhoneUnlock": return .awaitingPhoneUnlock
        case "running": return .running
        case "stopped": return .stopped
        case "failed:install": return .failed(VMFailure(phase: .install, reason: "x"))
        case "failed:run": return .failed(VMFailure(phase: .run, reason: "x"))
        default: throw XCTSkip("unknown state token '\(token)'")
        }
    }

    private func parseEvent(_ token: String, reason: String?) throws -> VMEvent {
        switch token {
        case "startInstall": return .startInstall
        case "installSucceeded": return .installSucceeded
        case "installFailed": return .installFailed(reason ?? "")
        case "powerOn": return .powerOn
        case "guestUnlocked": return .guestUnlocked
        case "powerOff": return .powerOff
        case "runtimeFailed": return .runtimeFailed(reason ?? "")
        default: throw XCTSkip("unknown event token '\(token)'")
        }
    }

    private func parseEffect(_ token: String) throws -> VMEffect {
        switch token {
        case "attachInstallerISO": return .attachInstallerISO
        case "detachInstallerISO": return .detachInstallerISO
        case "startVirtualMachine": return .startVirtualMachine
        case "stopVirtualMachine": return .stopVirtualMachine
        default: throw XCTSkip("unknown effect token '\(token)'")
        }
    }

    /// Compare ignoring the placeholder failure reason baked into start-state
    /// tokens; a transition INTO failed pins the real reason.
    private func assertStateMatches(_ expectedToken: String, reason expectedReason: String?,
                                    _ actual: VMState, _ msg: String) throws {
        let expected = try parseState(expectedToken)
        switch (expected, actual) {
        case (.failed(let e), .failed(let a)):
            XCTAssertEqual(e.phase, a.phase, msg)
            if let expectedReason { XCTAssertEqual(expectedReason, a.reason, msg) }
        default:
            XCTAssertEqual(expected, actual, msg)
        }
    }

    private let epoch: @Sendable () -> Date = { Date(timeIntervalSince1970: 0) }

    // MARK: - Lifecycle

    func testAllVectorTransitionsHold() throws {
        let vectors = try loadVectors()
        let lifecycle = vectors["lifecycle"] as? [String: Any]
        let transitions = lifecycle?["transitions"] as? [[String: Any]] ?? []
        var count = 0
        for t in transitions {
            let start = t["start"] as! String
            let ev = t["event"] as! String
            let reason = t["reason"] as? String
            let next = t["next"] as! String
            let effects = try (t["effects"] as! [String]).map { try parseEffect($0) }
            // No "sealed" key ⇒ the transition must hold for BOTH values.
            let sealedValues: [Bool] = (t["sealed"] as? Bool).map { [$0] } ?? [true, false]
            for sealedAtBoot in sealedValues {
                var lc = VMLifecycle(state: try parseState(start),
                                     sealedAtBoot: sealedAtBoot, clock: epoch)
                let got = try lc.handle(try parseEvent(ev, reason: reason))
                let label = "\(start) --\(ev)--> \(next) (sealed=\(sealedAtBoot))"
                try assertStateMatches(next, reason: reason, lc.state, label)
                XCTAssertEqual(effects, got, "effects for \(label)")
                count += 1
            }
        }
        XCTAssertGreaterThanOrEqual(count, 15, "vector file must actually contain transitions")
    }

    func testAllVectorInvalidTransitionsThrowAndLeaveStateUntouched() throws {
        let vectors = try loadVectors()
        let lifecycle = vectors["lifecycle"] as? [String: Any]
        let invalid = lifecycle?["invalid"] as? [[String: Any]] ?? []
        var count = 0
        for t in invalid {
            let start = t["start"] as! String
            let ev = t["event"] as! String
            for sealedAtBoot in [true, false] {
                let startState = try parseState(start)
                var lc = VMLifecycle(state: startState, sealedAtBoot: sealedAtBoot, clock: epoch)
                XCTAssertThrowsError(try lc.handle(try parseEvent(ev, reason: "r")),
                                     "\(start) --\(ev)--> must be rejected")
                // State must not change on a rejected event.
                XCTAssertEqual(lc.state, startState, "\(start) unchanged after rejected \(ev)")
                count += 1
            }
        }
        XCTAssertGreaterThanOrEqual(count, 20, "vector file must actually contain invalid cases")
    }

    // MARK: - Duration-gated install verdict

    func testInstallVerdictMatchesVectors() throws {
        let vectors = try loadVectors()
        let section = vectors["installVerdict"] as! [String: Any]
        XCTAssertEqual((section["minPlausibleInstallSeconds"] as! NSNumber).doubleValue,
                       VMLifecycle.minPlausibleInstallDuration)
        let start = Date(timeIntervalSince1970: 0)
        for c in section["cases"] as! [[String: Any]] {
            let elapsed = (c["elapsedSeconds"] as! NSNumber).doubleValue
            let expected = c["verdict"] as! String
            let verdict = VMLifecycle.verdictForCleanInstallStop(
                installStartedAt: start, now: start.addingTimeInterval(elapsed))
            switch verdict {
            case .installed:
                XCTAssertEqual(expected, "installSucceeded", "elapsed=\(elapsed)")
            case .failedTooFast:
                XCTAssertEqual(expected, "installFailed", "elapsed=\(elapsed)")
            }
        }
    }

    // MARK: - Resource plan

    func testResourcePlanMatchesVectors() throws {
        let vectors = try loadVectors()
        let section = vectors["resourcePlan"] as! [String: Any]
        for c in section["cases"] as! [[String: Any]] {
            let hostCpus = (c["hostCpus"] as! NSNumber).intValue
            let hostRamGiB = (c["hostRamGiB"] as! NSNumber).uint64Value
            let host = HostResources(cpuCount: hostCpus,
                                     memoryBytes: hostRamGiB * VMResourcePlan.gib)
            let label = "cpus=\(hostCpus) ramGiB=\(hostRamGiB)"
            XCTAssertEqual((c["vmCpus"] as! NSNumber).intValue,
                           VMResourcePlan.vmCPUCount(host: host), "vmCpus \(label)")
            XCTAssertEqual((c["vmMemGiB"] as! NSNumber).uint64Value * VMResourcePlan.gib,
                           VMResourcePlan.vmMemoryBytes(host: host), "vmMemGiB \(label)")
            XCTAssertEqual((c["maxVMs"] as! NSNumber).intValue,
                           VMResourcePlan.maxVMCount(host: host), "maxVMs \(label)")
        }
    }

    // MARK: - Bundle-name validation

    func testNameValidationMatchesVectors() throws {
        let vectors = try loadVectors()
        let section = vectors["nameValidation"] as! [String: Any]
        for n in section["valid"] as! [String] {
            XCTAssertTrue(VMInventoryStore.isValidName(n), "'\(n)' should be valid")
        }
        for n in section["invalid"] as! [String] {
            XCTAssertFalse(VMInventoryStore.isValidName(n), "'\(n)' should be invalid")
        }
    }
}

File: apps/builder-windows/tests/VMCoreVectorTests.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using Xunit;
using Flagship.Builder.VM;

namespace Flagship.Builder.Tests;

/// <summary>
/// Drives the pure VM core against the SHARED golden vectors
/// (apps/desktop-shared/golden/vm-core-vectors.json) — the cross-language
/// contract that keeps this C# core and the Mac Swift core identical, the way
/// engine/golden/preseed-vectors.json pins the preseed engine. If a vector
/// fails here, the fix is byte-parity with the vectors, never editing the
/// vectors to match the code (unless BOTH platforms change together).
/// </summary>
public class VMCoreVectorTests
{
    private static readonly JsonElement Vectors = LoadVectors();

    private static JsonElement LoadVectors()
    {
        var path = Path.Combine(AppContext.BaseDirectory, "Resources", "vm-core-vectors.json");
        var doc = JsonDocument.Parse(File.ReadAllBytes(path));
        return doc.RootElement;
    }

    // ---- token codecs ----

    private static VMState ParseState(string token) => token switch
    {
        "created" => VMState.Created,
        "installing" => VMState.Installing,
        "installed" => VMState.Installed,
        "awaitingPhoneUnlock" => VMState.AwaitingPhoneUnlock,
        "running" => VMState.Running,
        "stopped" => VMState.Stopped,
        "failed:install" => VMState.Failed(VMFailurePhase.Install, "x"),
        "failed:run" => VMState.Failed(VMFailurePhase.Run, "x"),
        _ => throw new ArgumentException($"unknown state token '{token}'"),
    };

    private static VMEvent ParseEvent(string token, string? reason) => token switch
    {
        "startInstall" => VMEvent.StartInstall,
        "installSucceeded" => VMEvent.InstallSucceeded,
        "installFailed" => VMEvent.InstallFailed(reason ?? ""),
        "powerOn" => VMEvent.PowerOn,
        "guestUnlocked" => VMEvent.GuestUnlocked,
        "powerOff" => VMEvent.PowerOff,
        "runtimeFailed" => VMEvent.RuntimeFailed(reason ?? ""),
        _ => throw new ArgumentException($"unknown event token '{token}'"),
    };

    private static VMEffect ParseEffect(string token) => token switch
    {
        "attachInstallerISO" => VMEffect.AttachInstallerISO,
        "detachInstallerISO" => VMEffect.DetachInstallerISO,
        "startVirtualMachine" => VMEffect.StartVirtualMachine,
        "stopVirtualMachine" => VMEffect.StopVirtualMachine,
        _ => throw new ArgumentException($"unknown effect token '{token}'"),
    };

    /// <summary>Compare ignoring the placeholder failure reason baked into
    /// start-state tokens; a transition INTO failed pins the real reason.</summary>
    private static void AssertStateMatches(string expectedToken, string? expectedReason, VMState actual)
    {
        var expected = ParseState(expectedToken);
        Assert.Equal(expected.Kind, actual.Kind);
        if (expected.Kind == VMStateKind.Failed)
        {
            Assert.Equal(expected.Failure!.Phase, actual.Failure!.Phase);
            if (expectedReason != null) Assert.Equal(expectedReason, actual.Failure!.Reason);
        }
    }

    // ---- lifecycle ----

    [Fact]
    public void AllVectorTransitionsHold()
    {
        int count = 0;
        foreach (var t in Vectors.GetProperty("lifecycle").GetProperty("transitions").EnumerateArray())
        {
            var start = t.GetProperty("start").GetString()!;
            var ev = t.GetProperty("event").GetString()!;
            var reason = t.TryGetProperty("reason", out var r) ? r.GetString() : null;
            var next = t.GetProperty("next").GetString()!;
            var effects = t.GetProperty("effects").EnumerateArray()
                           .Select(e => ParseEffect(e.GetString()!)).ToArray();
            // No "sealed" key ⇒ the transition must hold for BOTH values.
            bool[] sealedValues = t.TryGetProperty("sealed", out var s)
                ? new[] { s.GetBoolean() } : new[] { true, false };

            foreach (var sealedAtBoot in sealedValues)
            {
                var lc = new VMLifecycle(sealedAtBoot, ParseState(start),
                                         () => DateTimeOffset.UnixEpoch);
                var got = lc.Handle(ParseEvent(ev, reason));
                AssertStateMatches(next, reason, lc.State);
                Assert.Equal(effects, got);
                count++;
            }
        }
        Assert.True(count >= 15, "vector file must actually contain transitions");
    }

    [Fact]
    public void AllVectorInvalidTransitionsThrowAndLeaveStateUntouched()
    {
        int count = 0;
        foreach (var t in Vectors.GetProperty("lifecycle").GetProperty("invalid").EnumerateArray())
        {
            var start = t.GetProperty("start").GetString()!;
            var ev = t.GetProperty("event").GetString()!;
            foreach (var sealedAtBoot in new[] { true, false })
            {
                var startState = ParseState(start);
                var lc = new VMLifecycle(sealedAtBoot, startState, () => DateTimeOffset.UnixEpoch);
                Assert.Throws<VMLifecycleException>(() => lc.Handle(ParseEvent(ev, "r")));
                Assert.Equal(startState, lc.State); // state must not change on a rejected event
                count++;
            }
        }
        Assert.True(count >= 20, "vector file must actually contain invalid cases");
    }

    // ---- duration-gated install verdict ----

    [Fact]
    public void InstallVerdictMatchesVectors()
    {
        var section = Vectors.GetProperty("installVerdict");
        Assert.Equal(section.GetProperty("minPlausibleInstallSeconds").GetDouble(),
                     VMLifecycle.MinPlausibleInstallDuration.TotalSeconds);
        foreach (var c in section.GetProperty("cases").EnumerateArray())
        {
            var elapsed = TimeSpan.FromSeconds(c.GetProperty("elapsedSeconds").GetDouble());
            var expected = c.GetProperty("verdict").GetString()!;
            var verdict = VMLifecycle.VerdictForCleanInstallStop(elapsed);
            Assert.Equal(expected == "installSucceeded" ? VMEventKind.InstallSucceeded : VMEventKind.InstallFailed,
                         verdict.Kind);
            if (verdict.Kind == VMEventKind.InstallFailed)
                Assert.False(string.IsNullOrWhiteSpace(verdict.Reason), "failure verdicts carry an actionable reason");
        }
    }

    // ---- resource plan ----

    [Fact]
    public void ResourcePlanMatchesVectors()
    {
        foreach (var c in Vectors.GetProperty("resourcePlan").GetProperty("cases").EnumerateArray())
        {
            var host = new HostResources(
                c.GetProperty("hostCpus").GetInt32(),
                c.GetProperty("hostRamGiB").GetUInt64() * VMResourcePlan.GiB);
            var label = $"cpus={host.CpuCount} ramGiB={c.GetProperty("hostRamGiB").GetUInt64()}";
            Assert.True(c.GetProperty("vmCpus").GetInt32() == VMResourcePlan.VmCpuCount(host), $"vmCpus mismatch for {label}");
            Assert.True(c.GetProperty("vmMemGiB").GetUInt64() * VMResourcePlan.GiB == VMResourcePlan.VmMemoryBytes(host), $"vmMemGiB mismatch for {label}");
            Assert.True(c.GetProperty("maxVMs").GetInt32() == VMResourcePlan.MaxVMCount(host), $"maxVMs mismatch for {label}");
        }
    }

    // ---- bundle-name validation ----

    [Fact]
    public void NameValidationMatchesVectors()
    {
        var section = Vectors.GetProperty("nameValidation");
        foreach (var n in section.GetProperty("valid").EnumerateArray())
            Assert.True(VMInventoryStore.IsValidName(n.GetString()!), $"'{n.GetString()}' should be valid");
        foreach (var n in section.GetProperty("invalid").EnumerateArray())
            Assert.False(VMInventoryStore.IsValidName(n.GetString()!), $"'{n.GetString()}' should be invalid");
    }
}
